src/provider/provider_coarse.c

```diff
@@ -0,0 +1,1445 @@
+        LOG_ERR("unlocking the lock failed");
+        umf_result = UMF_RESULT_ERROR_UNKNOWN;
+        goto unlock_error;
+    }
+
+    return UMF_RESULT_SUCCESS;
+
+unlock_error:
+    ravl_tree_rm(coarse_provider->all_blocks, *resultPtr);
+block_error:
+    ravl_tree_rm(coarse_provider->upstream_alloc, *resultPtr);
+alloc_error:
+    if (!coarse_provider->disable_upstream_provider_free) {
+        umfMemoryProviderFree(coarse_provider->upstream_memory_provider,
+                              *resultPtr, size);
+    }
+    return umf_result;
+}
+
+static umf_result_t coarse_memory_provider_free(void *provider, void *ptr,
+                                                size_t bytes) {
+    if (provider == NULL) {
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    coarse_memory_provider_t *coarse_provider =
+        (struct coarse_memory_provider_t *)provider;
+
+    if (utils_mutex_lock(&coarse_provider->lock) != 0) {
+        LOG_ERR("lockng the lock failed");
+        return UMF_RESULT_ERROR_UNKNOWN;
+    }
+
+    assert(debug_check(coarse_provider));
+
+    block_t *block = ravl_tree_find(coarse_provider->all_blocks, ptr);
+    if (block == NULL) {
+        // the block was not found
+        LOG_ERR("memory block not found");
+        return UMF_RESULT_ERROR_UNKNOWN;
+    }
+
+    if (bytes) {
+        assert(bytes == block->size);
+    }
+
+    LOG_DEBUG("coarse_FREE (return_block_to_pool) %zu used %zu alloc %zu",
+              block->size, coarse_provider->used_size - block->size,
+              coarse_provider->alloc_size);
+
+    assert(coarse_provider->used_size >= block->size);
+    coarse_provider->used_size -= block->size;
+
+    block->used = false;
+
+    // Merge with prev and/or next block if they are unused and have continuous data.
+    block = free_block_merge_with_prev(
+        coarse_provider->upstream_memory_provider, coarse_provider->all_blocks,
+        coarse_provider->free_blocks, block);
+    block = free_block_merge_with_next(
+        coarse_provider->upstream_memory_provider, coarse_provider->all_blocks,
+        coarse_provider->free_blocks, block);
+
+    int rv = free_blocks_add(coarse_provider->free_blocks, block);
+    if (rv) {
+        return UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+    }
+
+    assert(debug_check(coarse_provider));
+
+    if (utils_mutex_unlock(&coarse_provider->lock) != 0) {
+        LOG_ERR("unlocking the lock failed");
+        return UMF_RESULT_ERROR_UNKNOWN;
+    }
+
+    return UMF_RESULT_SUCCESS;
+}
+
+static void coarse_memory_provider_get_last_native_error(void *provider,
+                                                         const char **ppMessage,
+                                                         int32_t *pError) {
+    (void)provider; // unused
+
+    if (ppMessage == NULL || pError == NULL) {
+        assert(0);
+        return;
+    }
+
+    // no UMF_RESULT_ERROR_MEMORY_PROVIDER_SPECIFIC error used
```

Comment:
is this a TODO comment...? or just a statement (if so, perhaps a short sentence why, or something)

Reply:
Done